### DISCUSSION TITLE
Why `Multiple logging methods cannot use the same event ID` (`SYSLIB1006`)?

### DISCUSSION THREAD

**Discussion by maxkoshevoi** [dotnet/runtime#70147] [Q&A]
SYSLIB1006 warning states that multiple logging methods cannot use the same event ID. But what's wrong with that?
https://docs.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib1006

I've annotated my methods without event id, so it's always `-1`
```cs
[LoggerMessage(Level = LogLevel.Information, Message = "Hi there")]
...
[LoggerMessage(Level = LogLevel.Information, Message = "Hi there2")]
```

and now I get a `SYSLIB1006` warning. 
![image](https://user-images.githubusercontent.com/6609929/171640795-c93cd333-e794-41c8-bcb7-7d0edfb23f79.png)

Is using event id mandatory? I don't use it in regular (not source generated) logs

**Comment by teo-tsirpanis** [comment]
They can, which is the reason this is a warning and not an error. It's just a bad practice.

**Reply by maxkoshevoi** [reply]
Yes, but when warning is a pretty serious event lever. Especially considering that a lot of code bases have `Treat warnings as errors` turned one (like ours).

Why is this a bad practice?

**Reply by teo-tsirpanis** [reply]
Disclaimer: I'm not 100% familiar in this area. Events are identified by their ID. In a log viewer if you want to search for occurences of a specific event it would be much easier to search for its ID. The message might change but the same event would have the same ID.

Maybe the warning should not fire on attributes that do not specify an ID, but having events without one seems a ~~bad~~ not-so-good practice as well.

**Reply by davidfowl** [reply]
IDs are also unique per category.

**Reply by maxkoshevoi** [reply]
Okay, I get that IDs are useful, but we don't use them in our codebase. Creating giant list (enum / class with consts) of possible IDs for every service doesn't seem to be worth it for now.

> Maybe the warning should not fire on attributes that do not specify an ID

Current logging interface provides overloads without EventID (that are not marked as obsolete and don't produce any other diagnostic event), so I would say the source generator's behavior should be consistent (not generate a warning if user haven't specified the ID)

**Reply by davidfowl** [reply]
OR you can disable the warning globally. The benefit of the source generator is that we can, at compile time try to recommend some best practices.

**Reply by maxkoshevoi** [reply]
Warning is a pretty harsh recommendation, since a lot of people have `Treat warnings as errors` enabled 😅
By default .Net generates static analysis warning only if there's some critical thing that needs to be addressed (from my experience).
Most "best-practice" warnings are ether in separate packages (like was with FxCop), so users would need to explicitly opt-in by installing them or got downgraded to suggestions, or turned off entirely (and again, users can opt-into them by setting [AnalysisLevel](https://devblogs.microsoft.com/dotnet/automatically-find-latent-bugs-in-your-code-with-net-5/)) 

Disabling it globally is not that straight-forward when you have 20+ solutions in 15 repositories..

**Reply by davidfowl** [reply]
That's true we should probably downgrade this to an info level message. Can you file an issue?

**Reply by maxkoshevoi** [reply]
As it stands 
- `You haven't provided an argument that matches a placeholder in the message` (which is a big issue that needs to be addressed), 
- `You've specified the same EventId for multiple logging methods` (also a big red flag)
- and `You haven't specified EventID for multiple log methods, and we recommend you doing so` (which is just a recommendation, everything will work as expected without fixing it)

have the same severity level (warning)

**Reply by maxkoshevoi** [reply]
> That's true we should probably downgrade this to an info level message. Can you file an issue?

Yes, thank you for understanding =)
Sorry, typed previous message before seeing your reply

**Reply by maxkoshevoi** [reply]
Created an issue: https://github.com/dotnet/runtime/issues/72263

**Comment by DerAlbertCom** [comment]
Maybe Microsoft should get the docs straight.

`An event ID associates a set of events. For example, all logs related to displaying a list of items on a page might be 1001.`
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-7.0#leid
